tensorflow_quantum/python/util.py

```diff
@@ -519,3 +519,95 @@ def exponential(operators, coefficients=None):
+        |1><1|_k = 0.5(I_k - Z_k).
+
+    Args:
+        qubit: A `cirq.GridQubit` on which the projector is supported.
+
+    Returns:
+        `cirq.PauliSum` representing the projector.
+    """
+    if not isinstance(qubit, cirq.GridQubit):
+        raise TypeError("A projector must live on a cirq.GridQubit.")
+    return 0.5 * cirq.I(qubit) - 0.5 * cirq.Z(qubit)
+
+
+def registers_from_precisions(precisions):
+    """Returns list of cirq.GridQubit registers for the given precisions.
+
+    Args:
+        precisions: a Python `list` of `int`s.  Entry `precisions[i]` sets
+            the number of qubits on which quantum integer `i` is supported.
+
+    Returns:
+        register_list: lists of `cirq.GridQubit`s, such that
+            len(register_list[i]) == precisions[i] and all entries are unique.
+    """
+    if not isinstance(precisions, list):
+        raise TypeError("Argument qubits must be a list of cirq.GridQubits.")
+    register_list = []
+    for r, width in enumerate(precisions):
+        this_register = []
+        for col in range(width):
+            this_register.append(cirq.GridQubit(r, col))
+        register_list.append(this_register)
+    return register_list
+
+
+def unsigned_integer_operator(qubits):
+    """Returns operator representing position in binary on a qubit register.
+
+    Unsigned integers on computers can be represented as a bitstring.  For an
+    integer represented by N bits, the k-th bit represents the presence
+    of the number 2**(N - k - 1) in the sum representing the integer.
+    Similarly, we can define a binary operator J as
+        J = \sum_P{k=0}^{N-1} 2^{N-k-1}|1><1|_k,
+    where
+        |1><1|_k = 0.5(I_k - Z_k).
+    J can be represented by a `cirq.PauliSum`.
+
+    Args:
+        qubits: Python `list` of `GridQubit`s on which the operator is
+            supported.
+
+    Returns:
+        int_op: A `cirq.PauliSum` representing the integer operator.
+    """
+    if not isinstance(qubits, list):
+        raise TypeError("Argument qubits must be a list of cirq.GridQubits.")
+    int_op = cirq.PauliSum()
+    width = len(qubits)
+    for loc, q in enumerate(qubits):
+        int_op += 2**(width - 1 - loc) * projector_on_one(q)
+    return int_op
+
+
+def unsigned_to_momentum_basis(register, circuit):
+    """Convert a circuit on qubits to the momentum basis.
+
+    Given a circuit C in the position basis, conversion to momentum basis
+    requires conjugation by the quantum Fourier transform Q, so that the
+    circuit in the momentum basis is --- Q --- C --- Q**-1 ---
+
+    Args:
+        qubits: a Python `list` of qubits on which the circuit is supported.
+        circuit: a `cirq.Circuit` to be taken to the momentum basis.
+
+    Returns:
+        `cirq.Circuit` which is the momentum basis version of the given circuit.
+    """
+    convert = cirq.ConvertToCzAndSingleGates(allow_partial_czs=True)
+    transform = cirq.Circuit(cirq.QFT(*register))
+    convert(transform)
```

Comment:
Can this just be `cirq.decompose` instead ?

Reply:
cirq.decompose left gates we do not support